File: php/spec/json.spec.php
<?php declare(strict_types=1);

use Enotype\Loaders;

describe('json', function() {
  describe('with \'{ "valid": true }\'', function() {
    it('produces the expected result', function() {
      expect(Loaders::json('{ "valid": true }'))->toEqual((object) [ 'valid' => true ]);
    });
  });
  
  describe('with \'42\'', function() {
    it('produces the expected result', function() {
      expect(Loaders::json('42'))->toBe(42);
    });
  });
  
  describe('with \'["valid", true]\'', function() {
    it('produces the expected result', function() {
      expect(Loaders::json('["valid", true]'))->toEqual(['valid', true]);
    });
  });
  
  describe('with \'invalid\'', function() {
    it('throws an exception', function() {
      expect(function() { Loaders::json('invalid'); })->toThrow();
    });
  });
  
  describe('with \'{ invalid: true }\'', function() {
    it('throws an exception', function() {
      expect(function() { Loaders::json('{ invalid: true }'); })->toThrow();
    });
  });
  
  describe('with \'{ "invalid": true, }\'', function() {
    it('throws an exception', function() {
      expect(function() { Loaders::json('{ "invalid": true, }'); })->toThrow();
    });
  });
});

File: ruby/spec/json.spec.rb
require 'enotype'

describe 'json' do
  context 'with \'{ "valid": true }\'' do
    it 'produces the expected result' do
      expect(Enotype.json('{ "valid": true }')).to eq({ 'valid' => true })
    end
  end
  
  context 'with \'42\'' do
    it 'produces the expected result' do
      expect(Enotype.json('42')).to be(42)
    end
  end
  
  context 'with \'["valid", true]\'' do
    it 'produces the expected result' do
      expect(Enotype.json('["valid", true]')).to eq(['valid', true])
    end
  end
  
  context 'with \'invalid\'' do
    it 'raises an error' do
      expect { Enotype.json('invalid') }.to raise_error(RuntimeError)
    end
  end
  
  context 'with \'{ invalid: true }\'' do
    it 'raises an error' do
      expect { Enotype.json('{ invalid: true }') }.to raise_error(RuntimeError)
    end
  end
  
  context 'with \'{ "invalid": true, }\'' do
    it 'raises an error' do
      expect { Enotype.json('{ "invalid": true, }') }.to raise_error(RuntimeError)
    end
  end
end